paddle/fluid/inference/tensorrt/plugin/deformable_conv_op_plugin.cu

```diff
@@ -0,0 +1,571 @@
+                            paddings_[0], strides_[0]);
+  ret.d[2] = ConvOutputSize(inputs[0].d[2], kernel_dims_[3], dilations_[1],
+                            paddings_[1], strides_[1]);
+  return ret;
+}
+
+bool DeformableConvPlugin::supportsFormat(nvinfer1::DataType type,
+                                          nvinfer1::TensorFormat format) const {
+  return ((type == data_type_ || type == nvinfer1::DataType::kINT32) &&
+          format == nvinfer1::TensorFormat::kLINEAR);
+}
+
+size_t DeformableConvPlugin::getWorkspaceSize(int max_batch_size) const {
+  int c_i = input_dim_[0], h_i = input_dim_[1], w_i = input_dim_[2];
+  int k_h = kernel_dims_[2], k_w = kernel_dims_[3];
+  int c_o = output_dim_[0], h_o = output_dim_[1], w_o = output_dim_[2];
+  int num_bytes = (data_type_ == nvinfer1::DataType::kFLOAT ? 4 : 2);
+  size_t data_col_size = static_cast<size_t>(c_i * k_h * k_w * im2col_step_ *
+                                             h_o * w_o * num_bytes);
+  return data_col_size;
+}
+
+int DeformableConvPlugin::enqueue(int batch_size, const void* const* inputs,
+#if IS_TRT_VERSION_LT(8000)
+                                  void** outputs, void* workspace,
+#else
+                                  void* const* outputs, void* workspace,
+#endif
+                                  cudaStream_t stream) {
+  if (data_type_ == nvinfer1::DataType::kFLOAT) {
+    return enqueue_impl<float>(batch_size, inputs, outputs, workspace, stream);
+  } else if (data_type_ == nvinfer1::DataType::kHALF) {
+    return enqueue_impl<half>(batch_size, inputs, outputs, workspace, stream);
+  }
+  assert("unsupported type.");
+}
+
+template <typename T>
+__device__ T kFloor(T x);
+
+template <>
+__device__ half kFloor<half>(half x) {
+  return hfloor(x);
+}
+
+template <>
+__device__ float kFloor<float>(float x) {
+  return floor(x);
+}
+
+template <typename T>
+__device__ T DmcnIm2colBilinear(const T* bottom_data, const int data_width,
+                                const int height, const int width, T h, T w) {
+  int h_low = kFloor<T>(h);
+  int w_low = kFloor<T>(w);
+  int h_high = h_low + 1;
+  int w_high = w_low + 1;
+
+  T lh = h - static_cast<T>(h_low);
+  T lw = w - static_cast<T>(w_low);
+  T hh = static_cast<T>(1) - lh, hw = static_cast<T>(1) - lw;
+
+  T v1 = 0;
+  if (h_low >= 0 && w_low >= 0) v1 = bottom_data[h_low * data_width + w_low];
+  T v2 = 0;
+  if (h_low >= 0 && w_high <= width - 1)
+    v2 = bottom_data[h_low * data_width + w_high];
+  T v3 = 0;
+  if (h_high <= height - 1 && w_low >= 0)
+    v3 = bottom_data[h_high * data_width + w_low];
+  T v4 = 0;
+  if (h_high <= height - 1 && w_high <= width - 1)
+    v4 = bottom_data[h_high * data_width + w_high];
+
+  T w1 = hh * hw, w2 = hh * lw, w3 = lh * hw, w4 = lh * lw;
+
+  T val = (w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4);
+  return val;
+}
+
+template <typename T>
+__global__ void ModulatedDeformableIm2colGpuKernel(
```

Comment:
the DmcnIm2colBilinear  and  ModulatedDeformableIm2colGpuKernel, ..  are duplicated with  https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/deformable_conv_op.cu

need to  refine


And, does the kernel support to fp16?  it is not supported in deformable_conv_op.cu

Reply:
fp16 is supported in dcn trt plugin but not supported in deformable_conv_op.cu